decryptor/postgresql/pg_decryptor.go

```diff
@@ -214,89 +229,108 @@ func (proxy *PgProxy) ProxyClientConnection(errCh chan<- error) {
-			} else {
-				logger.WithField("sql", queryWithHiddenValues).Debugln("New query")
-			}
-		}
+		censorSpan.End()
 
-		if censorErr := proxy.censor.HandleQuery(query); censorErr != nil {
-			censorSpan.End()
-			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorQueryIsNotAllowed).WithError(censorErr).Errorln("AcraCensor blocked query")
-			errorMessage, err := NewPgError("AcraCensor blocked this query")
+		// If the packet has been rejected by AcraCensor, stop here and don't send it to the database.
+		// Also, craft and send the client an error so that they know their query has been rejected.
+		if censored {
+			err := proxy.sendClientAcraCensorError(logger)
 			if err != nil {
-				logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCodingPostgresqlCantGenerateErrorPacket).WithError(err).Errorln("Can't create PostgreSQL error message")
-				errCh <- err
-				return
-			}
-			n, err := proxy.clientConnection.Write(errorMessage)
-			if err := base.CheckReadWrite(n, len(errorMessage), err); err != nil {
-				errCh <- err
-				return
-			}
-			n, err = proxy.clientConnection.Write(ReadyForQueryPacket)
-			if err := base.CheckReadWrite(n, len(ReadyForQueryPacket), err); err != nil {
 				errCh <- err
 				return
 			}
 			continue
 		}
 
-		newQuery, changed, err := proxy.queryObserverManager.OnQuery(base.NewOnQueryObjectFromQuery(query))
-		if err != nil {
-			logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorEncryptQueryData).Errorln("Error occurred on query handler")
+		// After tha packet has been observed and possibly modified, forward it to the database.
+		if err := packet.sendPacket(); err != nil {
+			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).
+				WithError(err).Errorln("Can't send packet")
+			errCh <- err
+			return
 		}
-		if changed {
-			packet.ReplaceQuery(newQuery.Query())
+		// If this is a termination packet, we're done here. Signal EOF and stop the proxy.
+		if packet.terminatePacket {
+			errCh <- io.EOF
+			return
 		}
+	}
+}
 
-		censorSpan.End()
+func (proxy *PgProxy) handleClientPacket(packet *PacketHandler, logger *log.Entry) (bool, error) {
+	// Let the protocol observer take a look at the packet, keeping note of it.
+	packetType, err := proxy.protocolState.HandleClientPacket(packet)
+	if err != nil {
+		return false, err
+	}
+	switch packetType {
+	case QueryPacket:
+		// If that's some sort of a packet with query data inside it, go on...
 
-		if err := packet.sendPacket(); err != nil {
-			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).WithError(err).Errorln("Can't send packet")
-			errCh <- err
-			return
+	default:
+		// Forward all other uninteresting packets to the database without processing.
+		return false, nil
+	}
+	query := proxy.protocolState.PendingQuery()
```

Comment:
imho better to move it to separate function `handleQueryPacket` and call it in `case QueryPacket`. Because in a future we will handle `execute` and other packets and probably will be handled in a separate case.

Reply:
Mmm... Yeah, makes sense. I intended to keep the handling of other query-like packet similar, but it very much makes sense to have processing as

```go
switch type {
case This:
        handleThis()
case That:
        handleThat()
default:
        handleElse()
}
```
rather than the current
```go
switch type {
case This:
        // break out of switch
default:
        handleElse()
}
handleThis()
```